ussomacaroon/client_test.go

```diff
@@ -0,0 +1,212 @@
+	c.Assert(ms, gc.IsNil)
+}
+
+func (s *clientSuite) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.URL.Path {
+	case "/macaroon":
+		s.serveMacaroon(w, r)
+	case "/login":
+		s.serveLogin(w, r)
+	case "/api/v2/tokens/discharge":
+		s.serveDischarge(w, r)
+	default:
+		http.NotFound(w, r)
+	}
+}
+
+func (s *clientSuite) serveMacaroon(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		fail(w, r, errgo.Newf("bad method: %s", r.Method))
+	}
+	if s.macaroon != nil {
+		httprequest.WriteJSON(w, http.StatusOK, ussomacaroon.MacaroonResponse{
+			Macaroon: s.macaroon,
+		})
+	} else {
+		httprequest.WriteJSON(w, http.StatusInternalServerError, params.Error{
+			Message: "test error",
+		})
+	}
+}
+
+func (s *clientSuite) serveLogin(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		fail(w, r, errgo.Newf("bad method: %s", r.Method))
+	}
+	var lr ussomacaroon.LoginRequest
+	if err := httprequest.Unmarshal(httprequest.Params{Request: r, Response: w}, &lr); err != nil {
+		fail(w, r, err)
+	}
+	if n := len(lr.Macaroons.Macaroons); n != 1 {
+		fail(w, r, errgo.Newf("macaroon slice has unexpected length %d", n))
+	}
+	if id := lr.Macaroons.Macaroons[0].Id(); string(id) != "test macaroon" {
+		fail(w, r, errgo.Newf("unexpected macaroon sent %q", string(id)))
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (s *clientSuite) serveDischarge(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		fail(w, r, errgo.Newf("bad method: %s", r.Method))
+	}
+	var dr ussomacaroon.USSODischargeRequest
+	if err := httprequest.Unmarshal(httprequest.Params{Request: r, Response: w}, &dr); err != nil {
+		fail(w, r, err)
+	}
+	if dr.Discharge.Email == "" {
+		fail(w, r, errgo.New("email not specified"))
+	}
+	if dr.Discharge.Password == "" {
+		fail(w, r, errgo.New("password not specified"))
+	}
+	if dr.Discharge.OTP == "" {
+		fail(w, r, errgo.New("otp not specified"))
+	}
+	if dr.Discharge.CaveatID == "" {
+		fail(w, r, errgo.New("caveat_id not specified"))
+	}
+	if dr.Discharge.Email != "user@example.com" || dr.Discharge.Password != "secret" {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(`{"error_list": [{"message": "Provided email/password is not correct.", "code": "invalid-credentials"}], "message": "Provided email/password is not correct.", "code": "INVALID_CREDENTIALS", "extra": {}}`))
+		return
+	}
+	var m ussomacaroon.USSOMacaroon
+	m.Macaroon = *testDischargeMacaroon
+	httprequest.WriteJSON(w, http.StatusOK, map[string]interface{}{"discharge_macaroon": &m})
+}
+
+func fail(w http.ResponseWriter, r *http.Request, err error) {
+	httprequest.WriteJSON(w, http.StatusBadRequest, params.Error{
+		Message: err.Error(),
+	})
+}
+
+var testCause = errgo.New("test cause")
+var testMacaroon *macaroon.Macaroon
+var testDischargeMacaroon *macaroon.Macaroon
+
+func init() {
```

Comment:
why do these need to be in an init like this? Why not just put them in the suite?